src/transformers/modeling_dpr.py

```diff
@@ -0,0 +1,551 @@
+        start_logits = start_logits.squeeze(-1)
+        end_logits = end_logits.squeeze(-1)
+        rank_logits = self.qa_classifier(sequence_output[:, 0, :])
+        return start_logits, end_logits, rank_logits
+
+
+##############
+# Providers
+##############
+
+
+def get_bert_question_encoder_component(config: DprConfig, **kwargs) -> DprBertEncoder:
+    question_encoder = DprBertEncoder.init_encoder(
+        config.pretrained_model_cfg, projection_dim=config.projection_dim, **kwargs
+    )
+    return question_encoder
+
+
+def get_bert_ctx_encoder_component(config: DprConfig, **kwargs) -> DprBertEncoder:
+    ctx_encoder = DprBertEncoder.init_encoder(
+        config.pretrained_model_cfg, projection_dim=config.projection_dim, **kwargs
+    )
+    return ctx_encoder
+
+
+def get_bert_reader_component(config: DprConfig, **kwargs) -> Reader:
+    encoder = DprBertEncoder.init_encoder(config.pretrained_model_cfg, projection_dim=config.projection_dim, **kwargs)
+    hidden_size = encoder.config.hidden_size
+    reader = Reader(encoder, hidden_size)
+    return reader
+
+
+##################
+# PreTrainedModel
+##################
+
+
+CheckpointState = collections.namedtuple(
+    "CheckpointState", ["model_dict", "optimizer_dict", "scheduler_dict", "offset", "epoch", "encoder_params"]
+)
+
+
+def load_states_from_checkpoint(model_file: str) -> CheckpointState:
+    logger.info("Reading saved model from %s", model_file)
+    state_dict = torch.load(model_file, map_location=lambda s, l: default_restore_location(s, "cpu"))
+    logger.info("model_state_dict keys %s", state_dict.keys())
+    return CheckpointState(**state_dict)
+
+
+class DprPretrainedContextEncoder(PreTrainedModel):
+    """ An abstract class to handle weights initialization and
+        a simple interface for downloading and loading pretrained models.
+    """
+
+    config_class = DprConfig
+    load_tf_weights = None
+    base_model_prefix = "dpr"
+
+    def init_weights(self):
+        """Load the weights from the official DPR repo's format if specified."""
+        if self.config.biencoder_model_file is not None:
+            logger.info("Loading DPR biencoder from {}".format(self.config.biencoder_model_file))
+            saved_state = load_states_from_checkpoint(self.config.biencoder_model_file)
+            encoder, prefix = self.ctx_encoder, "ctx_model."
+            prefix_len = len(prefix)
+            ctx_state = {
+                key[prefix_len:]: value for (key, value) in saved_state.model_dict.items() if key.startswith(prefix)
+            }
+            encoder.load_state_dict(ctx_state)
+        else:
+            self.ctx_encoder.init_weights()
+
+
+class DprPretrainedQuestionEncoder(PreTrainedModel):
+    """ An abstract class to handle weights initialization and
+        a simple interface for downloading and loading pretrained models.
+    """
+
+    config_class = DprConfig
+    load_tf_weights = None
+    base_model_prefix = "dpr"
```

Comment:
not sure whether it's a good idea to have to different models have the same `base_model_prefix`. What do you think @LysandreJik ?

Reply:
Well `base_model_prefix` is here to indicate that the base transformer model of that model is named like the value of the variable. Here it seems none of the models have the `self.dpr` attribute as base model

Reply:
I guess I have to replace it by `question_encoder` (as in `self.question_encoder`) for this one then ? What is it used for ?